src/command/dev/watcher.rs

```diff
@@ -191,30 +204,49 @@ impl SubgraphSchemaWatcher {
                 match last_message {
                     Some(last_message) => {
                         if &subgraph_definition.sdl != last_message {
+                            if self.subgraph_retry_countdown < self.subgraph_retries {
+                                eprintln!(
+                                    "{} subgraph connectivity restored for {}",
+                                    Emoji::Reload,
+                                    self.subgraph_key.0
+                                )
+                            }
                             self.message_sender.update_subgraph(&subgraph_definition)?;
                         }
                     }
                     None => {
                         self.message_sender.add_subgraph(&subgraph_definition)?;
                     }
                 }
+                self.subgraph_retry_countdown = self.subgraph_retries;
                 Some(subgraph_definition.sdl)
             }
             Err(e) => {
-                let error_str = e.to_string();
-                match last_message {
-                    Some(prev_message) => {
-                        if &error_str != prev_message {
-                            print_error(e);
-                            self.message_sender.remove_subgraph(&self.subgraph_key.0)?;
-                        }
-                    }
-                    None => {
-                        print_error(e);
-                        let _ = self.message_sender.remove_subgraph(&self.subgraph_key.0);
-                    }
+                // `subgraph-retries` can be set by the user away from the default value of 0,
+                // this defaults to Rover's current behaviour.
+                //
+                // If a user does set this value to a non-zero one, and we get a non-retryable error
+                // from one of our subgraphs, we'll retain the old schema we had and continue
+                // operation. This will happen until the countdown hits 0 at which point the
+                // subgraph will be disconnected from the supergraph.
+                //
+                // Every time we successfully communicate with the subgraph we set the countdown
+                // back to the maximum value.
+                //
+                if self.subgraph_retry_countdown > 0 {
+                    self.subgraph_retry_countdown -= 1;
+                    eprintln!("{} error detected communicating with subgraph '{}', schema changes will not be reflected.\nWill retry but subgraph logs should be inspected", Emoji::Warn, &self.subgraph_key.0);
```

Comment:
to confirm, schema changes won't be saved because introspection couldn't run or because rover couldn't confirm the schema with the subgraph?

Reply:
So more likely the former, essentially this happens when the introspection query fails for whatever reason (4xx, 5xx and so on), so we don't get back a new schema to check against.